### PR TITLE
Call sampler on local child spans.

### LINKED ISSUE #1228: Attributes from Samplers not being added to sampled Spans
The [sampling section of the Tracing SDK specification](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#sampling) specifies that the return value of the ShouldSample method contains "A set of span Attributes that will also be added to the Span.". 

When using a custom sampler with the Go SDK, attributes returned from the Sampler are only set on the root Span in a Trace, not on each of the individual spans. I tested with the Python SDK, and in that implementation, Attributes are being set on every span. 

The use case we need this for is to be able to set the Sample Rate for each span sent to Honeycomb. Our backend uses the sample rate when calculating aggregate functions on events. Here's the [sampler implementation we're using](https://github.com/honeycombio/opentelemetry-samplers-go/blob/main/honeycombsamplers/deterministic_sampler.go#L58) for a proof of concept. 

Using [this sample app](https://github.com/honeycombio/otel-sampling-demos/blob/master/golang/deterministic/main.go), we get the following output:

```
[
  {
    "SpanContext": {
    ...
    },
    "Attributes": null,
  }
]
```

On each of the Spans, except to the root span:

```
[
  {
     "SpanContext": {
      ...
  },
  "Attributes": [
    {
      "Key": "SampleRate",
      "Value": {
        "Type": "INT32",
        "Value": 2
      }
    },
    ...
    ]
  }
]
```

Using the [Python sample app](https://github.com/honeycombio/otel-sampling-demos/blob/master/python/deterministic/app.py), I'm able to see the Attribute on every span:

```
    ...
    "attributes": {
        "SampleRate": 2
    },
    ...
```

### PR DESCRIPTION
**Pull request by paulosman**
Fixes #1228 

Previously, the sampler would only be called for the first span in a process. This makes sense for trace level sampling but doesn't allow the sampler to specify attributes to be set on each span, which is important for some implementations. This change removes the guard that limits sampling decisions to local root spans.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -24,6 +24,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
    They no longer track the gRPC codes. (#1214)
 - The `StatusCode` field of the `SpanData` struct in the `go.opentelemetry.io/otel/sdk/export/trace` package now uses the codes package from this package instead of the gRPC project. (#1214)
 - Move the `go.opentelemetry.io/otel/api/baggage` package into `go.opentelemetry.io/otel/propagators`. (#1217)
+- The Sampler is now called on local root spans. (#1228)
```
Comment (Aneurysm9) [inline CHANGELOG.md:27]:
```suggestion
- The Sampler is now called on local child spans. (#1228)
```

Comment (MrAlias) [inline CHANGELOG.md:27]:
```suggestion
- The Sampler is now called on local root spans. (#1233)
```

**Review by MrAlias — APPROVED** [review]

**Review by Aneurysm9 — APPROVED** [review]

**Review by XSAM — APPROVED** [review]

**Comment by MrAlias** [comment]
@paulosman it looks like maintainer edits are turned off for this PR so I can't merge master into this. If you can do this sync this PR looks ready to merge.

File: CHANGELOG.md [hunk 2]
```diff
@@ -36,6 +36,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
    They no longer track the gRPC codes. (#1214)
 - The `StatusCode` field of the `SpanData` struct in the `go.opentelemetry.io/otel/sdk/export/trace` package now uses the codes package from this package instead of the gRPC project. (#1214)
 - Move the `go.opentelemetry.io/otel/api/baggage` package into `go.opentelemetry.io/otel/propagators`. (#1217)
+- The Sampler is now called on local child spans. (#1233)
```
Comment (MrAlias) [inline CHANGELOG.md:39]: We'll need to move this up to the `Unreleased` section.

File: CHANGELOG.md [hunk 3]
```diff
@@ -8,6 +8,8 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 ## [Unreleased]
 
+- The Sampler is now called on local child spans. (#1233)
```
Comment (lizthegrey) [inline CHANGELOG.md:11]: this goes under "changed" in "unreleased".

**Review by lizthegrey — CHANGES_REQUESTED** [review]
please write tests for this.

**Comment by paulosman** [comment]
> please write tests for this.

Added a test to catch local child spans. Build is fixed, but circle seems stuck - pushed a changed and it won't rebuild.

**Review by lizthegrey — APPROVED** [review]

**Pull request merged**